# microbiome-in-DS

This repository contains the code allowing to reproduce the results of the analysis of the microbiome data, pre-published in the article: doi: https://doi.org/10.1101/2020.11.16.385989 Those data were obtained to determine the microbial content of the faecal microbiome samples from Wild Type and Down Syndrome model mice fed high-fat diet.

## What is in the package?
- script to run seed-kraken
- python code to process the .csv files
- R code to analyse the data and visualise results (analysis.R)

The .csv files with microbial species counts generated by seed-kraken tool are available here:
http://regulomics.mimuw.edu.pl/~ilona/microbiome_in_DS/

## Usage
Run: 
```analysis.R``` 
in RStudio to investigate all the steps of the analysis.

### Content:
# Figures:
 * Differentially abundant species (DA) - DeSeq2
 * Plotting the top DA species 
 * Plotting the most abundant species in DS and in WT mice
 * Venn diagrams of species present in WT and DS
 * Multidimensional Scaling Analysis (MDS) 
 * Hierarchical Clustering (HCL)
 * Correlation of species within WT and DS mice
 * Validation with total RNA HiSeq data - presence of species - Venn diagram
 * Correlation of species abundances in 16S and total RNA HiSeq data

### Supplementary Data and Figures:
 * Numbers of reads per sample in 16S data
 * Numbers of reads per sample in total RNA data 
 * Rarefaction curves in 16S data 
 * Rarefaction curves in total RNA data 
 * Validation with total RNA HiSeq data - presence of species - Venn diagram for each condition
 * Correlation of species within WT and DS mice - validation in total RNA data

